src/HazelcastClient.ts

```diff
@@ -133,21 +133,86 @@ export class HazelcastClient {
         this.loggingService = new LoggingService(this.config.customLogger,
             this.config.properties['hazelcast.logging.level'] as string);
         this.loadBalancer = this.initLoadBalancer();
-        this.listenerService = new ListenerService(this);
+        this.errorFactory = new ClientErrorFactory();
+
+        this.listenerService = new ListenerService(
+            this.getLoggingService().getLogger(),
+            this.getConfig().network.smartRouting,
+            () => this.getConnectionManager(),
```

Comment:
What's the point in passing closures instead of concrete objects? Service instances should be passed directly where possible.